### ISSUE TITLE
Deadlock on startup due to cannot aquire semaphore

### ISSUE THREAD

**Issue by Bananeweizen** [eclipse-equinox/equinox#84]
We have seen it multiple times (but still seldomly) that Eclipse hangs shortly after showing the splash. None of the typical fixes (-clean, -refresh, delete workspace/metadata/plugins/core.resources) changes anything. Using jstack we find that `EclipseStarter` basically runs an endless loop if it cannot aquire the given semaphore:

```
"main" #1 prio=5 os_prio=0 cpu=1296.88ms elapsed=57.81s allocated=59720K defined_classes=1010 tid=0x000002160ce55830 nid=0x156c runnable  [0x00000040d61be000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@17.0.3/Native Method)
	- parking to wait for  <0x00000007011ef200> (a java.util.concurrent.Semaphore$NonfairSync)
	at java.util.concurrent.locks.LockSupport.parkNanos(java.base@17.0.3/LockSupport.java:252)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@17.0.3/AbstractQueuedSynchronizer.java:717)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(java.base@17.0.3/AbstractQueuedSynchronizer.java:1074)
	at java.util.concurrent.Semaphore.tryAcquire(java.base@17.0.3/Semaphore.java:415)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.updateSplash(EclipseStarter.java:1191)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.setStartLevel(EclipseStarter.java:1146)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.startup(EclipseStarter.java:346)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:251)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@17.0.3/Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@17.0.3/NativeMethodAccessorImpl.java:77)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@17.0.3/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java.base@17.0.3/Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
```

The same semaphore cannot be found at other places of the full jstack output. It's also not clear under which circumstances this gets triggered. We have seen it both in 2022-06 and 2022-09 builds.

Looking at the code I have no idea what the semaphore actually protects. But wouldn't it be useful to leave the `while(true)` loop in `EclipseStarter.updateSplash()` after a certain time (that is, basically a fixed number of max iterations), in case the semaphore cannot be aquired at all?

**Comment by HannesWell** [comment]
I have encountered such very long start ups recently too.
And when using Visual-VM to profile the Eclipse instance the semaphore was waiting for the resolver threads to complete, which were heavily computing 'package-usages'. Can you try that?

IIRC from discussions in Tycho the Equinox resolver sometimes needs very long to check package-usages from the Manifest and because we are moving to more and more third-party bundles from Maven-Central, which often use BND to generate their Manifest, which by default adds 'uses' directives, more of usages are to consider/compute.

But I think @tjwatson, @mickaelistria and @laeubi know better about that topic (I only noticed this from the sideline).

**Comment by laeubi** [comment]
There are two parameter one can try out (see https://github.com/eclipse/tycho/issues/213#issuecomment-912547700):

- `equinox.resolver.revision.batch.size`
- `equinox.resolver.batch.timeout`

> which by default adds 'uses' directives, more of usages are to consider/compute
I found that actually heavy dependency-chains with require bundle give the resolver a hard time...

**Comment by Bananeweizen** [comment]
@HannesWell Yes, it's the same issue, CPU is burnt around `computeUses()`. I also took a memory snapshot, so someone with more understanding of the matter might be able to draw conclusions on whether the number of object instances from the resolver namespace is reasonable or not:
<img width="544" alt="image" src="https://user-images.githubusercontent.com/406876/180186653-27527959-00eb-4bc0-ad56-f69276eb0ada.png">

I've added the batch.size argument now and that fixed it. I now also remembered that we already set the timeout argument in our (rather big) product months/years ago. I still have to do some more experiments with the 2 arguments, but I will probably add at least one to our company Oomph setups.

I now have a workaround. Question is: Shall we still do something for the IDE? E.g. add this also to the EPP downloads? Otherwise I'm fine with closing.
